In [1]:
import pandas as pd
df=pd.read_csv("Cars_Data.csv")
df

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[432 rows x 15 columns]

###  Find all Null Values in the dataset. If there is any null value in any column, then fill it with the mean of that column.

In [2]:
df.head(5)

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [3]:
df.describe()

EngineSize   Cylinders  Horsepower    MPG_City  MPG_Highway  \
count  428.000000  426.000000  428.000000  428.000000   428.000000   
mean     3.196729    5.807512  215.885514   20.060748    26.843458   
std      1.108595    1.558443   71.836032    5.238218     5.741201   
min      1.300000    3.000000   73.000000   10.000000    12.000000   
25%      2.375000    4.000000  165.000000   17.000000    24.000000   
50%      3.000000    6.000000  210.000000   19.000000    26.000000   
75%      3.900000    6.000000  255.000000   21.250000    29.000000   
max      8.300000   12.000000  500.000000   60.000000    66.000000   

            Weight   Wheelbase      Length  
count   428.000000  428.000000  428.000000  
mean   3577.953271  108.154206  186.362150  
std     758.983215    8.311813   14.357991  
min    1850.000000   89.000000  143.000000  
25%    3104.000000  103.000000  178.000000  
50%    3474.500000  107.000000  187.000000  
75%    3977.750000  112.000000  194.000000  
max    7190.000000  144.000000  238.000000

In [4]:
df.info()
# So there is not any null value
# but there are some object data types \
# due to object data type some time .info() ignore na values
# so we use isnull() function

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 432 entries, 0 to 431
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    object 
 6   Invoice      428 non-null    object 
 7   EngineSize   428 non-null    float64
 8   Cylinders    426 non-null    float64
 9   Horsepower   428 non-null    float64
 10  MPG_City     428 non-null    float64
 11  MPG_Highway  428 non-null    float64
 12  Weight       428 non-null    float64
 13  Wheelbase    428 non-null    float64
 14  Length       428 non-null    float64
dtypes: float64(8), object(7)
memory usage: 50.8+ KB


In [5]:
df.isnull().sum()
# it shows us the null values

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      6
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

In [6]:
df.isna().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      6
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

In [7]:
df.mean()

EngineSize        3.196729
Cylinders         5.807512
Horsepower      215.885514
MPG_City         20.060748
MPG_Highway      26.843458
Weight         3577.953271
Wheelbase       108.154206
Length          186.362150
dtype: float64

In [8]:
df.mean()

EngineSize        3.196729
Cylinders         5.807512
Horsepower      215.885514
MPG_City         20.060748
MPG_Highway      26.843458
Weight         3577.953271
Wheelbase       108.154206
Length          186.362150
dtype: float64

In [9]:
df=df.fillna(df.mean())
df

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[432 rows x 15 columns]

In [10]:
df.loc[30]

Make                   NaN
Model                  NaN
Type                   NaN
Origin                 NaN
DriveTrain             NaN
MSRP                   NaN
Invoice                NaN
EngineSize        3.196729
Cylinders         5.807512
Horsepower      215.885514
MPG_City         20.060748
MPG_Highway      26.843458
Weight         3577.953271
Wheelbase       108.154206
Length           186.36215
Name: 30, dtype: object

In [11]:
df.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

Now only the 'object' type coloums are left so we remove  the rows and re set coloum index

In [12]:
df=df.dropna()

In [13]:
df.head(45)

Make                             Model    Type  Origin DriveTrain  \
0   Acura                               MDX     SUV    Asia        All   
1   Acura                    RSX Type S 2dr   Sedan    Asia      Front   
2   Acura                           TSX 4dr   Sedan    Asia      Front   
3   Acura                            TL 4dr   Sedan    Asia      Front   
4   Acura                        3.5 RL 4dr   Sedan    Asia      Front   
5   Acura           3.5 RL w/Navigation 4dr   Sedan    Asia      Front   
6   Acura            NSX coupe 2dr manual S  Sports    Asia       Rear   
7    Audi                       A4 1.8T 4dr   Sedan  Europe      Front   
8    Audi            A41.8T convertible 2dr   Sedan  Europe      Front   
9    Audi                        A4 3.0 4dr   Sedan  Europe      Front   
10   Audi         A4 3.0 Quattro 4dr manual   Sedan  Europe        All   
11   Audi           A4 3.0 Quattro 4dr auto   Sedan  Europe        All   
12   Audi                        A6 3.0 4dr   Sedan  Europe      Front   
13   Audi                A6 3.0 Quattro 4dr   Sedan  Europe        All   
14   Audi            A4 3.0 convertible 2dr   Sedan  Europe      Front   
15   Audi    A4 3.0 Quattro convertible 2dr   Sedan  Europe        All   
16   Audi          A6 2.7 Turbo Quattro 4dr   Sedan  Europe        All   
17   Audi                A6 4.2 Quattro 4dr   Sedan  Europe        All   
18   Audi                  A8 L Quattro 4dr   Sedan  Europe        All   
19   Audi                    S4 Quattro 4dr   Sedan  Europe        All   
20   Audi                          RS 6 4dr  Sports  Europe      Front   
21   Audi    TT 1.8 convertible 2dr (coupe)  Sports  Europe      Front   
22   Audi  TT 1.8 Quattro 2dr (convertible)  Sports  Europe        All   
23   Audi    TT 3.2 coupe 2dr (convertible)  Sports  Europe        All   
24   Audi              A6 3.0 Avant Quattro   Wagon  Europe        All   
25   Audi                  S4 Avant Quattro   Wagon  Europe        All   
26    BMW                           X3 3.0i     SUV  Europe        All   
27    BMW                           X5 4.4i     SUV  Europe        All   
28    BMW                          325i 4dr   Sedan  Europe       Rear   
29    BMW                         325Ci 2dr   Sedan  Europe       Rear   
31    BMW             325Ci convertible 2dr   Sedan  Europe       Rear   
32    BMW                         325xi 4dr   Sedan  Europe        All   
33    BMW                          330i 4dr   Sedan  Europe       Rear   
34    BMW                         330Ci 2dr   Sedan  Europe       Rear   
35    BMW                         330xi 4dr   Sedan  Europe        All   
36    BMW                          525i 4dr   Sedan  Europe       Rear   
37    BMW             330Ci convertible 2dr   Sedan  Europe       Rear   
38    BMW                          530i 4dr   Sedan  Europe       Rear   
40    BMW                         545iA 4dr   Sedan  Europe       Rear   
41    BMW                          745i 4dr   Sedan  Europe       Rear   
42    BMW                         745Li 4dr   Sedan  Europe       Rear   
43    BMW                      M3 coupe 2dr  Sports  Europe       Rear   
44    BMW                M3 convertible 2dr  Sports  Europe       Rear   
45    BMW           Z4 convertible 2.5i 2dr  Sports  Europe       Rear   
46    BMW           Z4 convertible 3.0i 2dr  Sports  Europe       Rear   

        MSRP   Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
0   $36,945   $33,337          3.5        6.0       265.0      17.0   
1   $23,820   $21,761          2.0        4.0       200.0      24.0   
2   $26,990   $24,647          2.4        4.0       200.0      22.0   
3   $33,195   $30,299          3.2        6.0       270.0      20.0   
4   $43,755   $39,014          3.5        6.0       225.0      18.0   
5   $46,100   $41,100          3.5        6.0       225.0      18.0   
6   $89,765   $79,978          3.2        6.0       290.0      17.0   
7   $25,940   $23,508          1

In [14]:
df = df.reset_index(drop=True)
df.head(45)

Make                             Model    Type  Origin DriveTrain  \
0   Acura                               MDX     SUV    Asia        All   
1   Acura                    RSX Type S 2dr   Sedan    Asia      Front   
2   Acura                           TSX 4dr   Sedan    Asia      Front   
3   Acura                            TL 4dr   Sedan    Asia      Front   
4   Acura                        3.5 RL 4dr   Sedan    Asia      Front   
5   Acura           3.5 RL w/Navigation 4dr   Sedan    Asia      Front   
6   Acura            NSX coupe 2dr manual S  Sports    Asia       Rear   
7    Audi                       A4 1.8T 4dr   Sedan  Europe      Front   
8    Audi            A41.8T convertible 2dr   Sedan  Europe      Front   
9    Audi                        A4 3.0 4dr   Sedan  Europe      Front   
10   Audi         A4 3.0 Quattro 4dr manual   Sedan  Europe        All   
11   Audi           A4 3.0 Quattro 4dr auto   Sedan  Europe        All   
12   Audi                        A6 3.0 4dr   Sedan  Europe      Front   
13   Audi                A6 3.0 Quattro 4dr   Sedan  Europe        All   
14   Audi            A4 3.0 convertible 2dr   Sedan  Europe      Front   
15   Audi    A4 3.0 Quattro convertible 2dr   Sedan  Europe        All   
16   Audi          A6 2.7 Turbo Quattro 4dr   Sedan  Europe        All   
17   Audi                A6 4.2 Quattro 4dr   Sedan  Europe        All   
18   Audi                  A8 L Quattro 4dr   Sedan  Europe        All   
19   Audi                    S4 Quattro 4dr   Sedan  Europe        All   
20   Audi                          RS 6 4dr  Sports  Europe      Front   
21   Audi    TT 1.8 convertible 2dr (coupe)  Sports  Europe      Front   
22   Audi  TT 1.8 Quattro 2dr (convertible)  Sports  Europe        All   
23   Audi    TT 3.2 coupe 2dr (convertible)  Sports  Europe        All   
24   Audi              A6 3.0 Avant Quattro   Wagon  Europe        All   
25   Audi                  S4 Avant Quattro   Wagon  Europe        All   
26    BMW                           X3 3.0i     SUV  Europe        All   
27    BMW                           X5 4.4i     SUV  Europe        All   
28    BMW                          325i 4dr   Sedan  Europe       Rear   
29    BMW                         325Ci 2dr   Sedan  Europe       Rear   
30    BMW             325Ci convertible 2dr   Sedan  Europe       Rear   
31    BMW                         325xi 4dr   Sedan  Europe        All   
32    BMW                          330i 4dr   Sedan  Europe       Rear   
33    BMW                         330Ci 2dr   Sedan  Europe       Rear   
34    BMW                         330xi 4dr   Sedan  Europe        All   
35    BMW                          525i 4dr   Sedan  Europe       Rear   
36    BMW             330Ci convertible 2dr   Sedan  Europe       Rear   
37    BMW                          530i 4dr   Sedan  Europe       Rear   
38    BMW                         545iA 4dr   Sedan  Europe       Rear   
39    BMW                          745i 4dr   Sedan  Europe       Rear   
40    BMW                         745Li 4dr   Sedan  Europe       Rear   
41    BMW                      M3 coupe 2dr  Sports  Europe       Rear   
42    BMW                M3 convertible 2dr  Sports  Europe       Rear   
43    BMW           Z4 convertible 2.5i 2dr  Sports  Europe       Rear   
44    BMW           Z4 convertible 3.0i 2dr  Sports  Europe       Rear   

        MSRP   Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
0   $36,945   $33,337          3.5        6.0       265.0      17.0   
1   $23,820   $21,761          2.0        4.0       200.0      24.0   
2   $26,990   $24,647          2.4        4.0       200.0      22.0   
3   $33,195   $30,299          3.2        6.0       270.0      20.0   
4   $43,755   $39,014          3.5        6.0       225.0      18.0   
5   $46,100   $41,100          3.5        6.0       225.0      18.0   
6   $89,765   $79,978          3.2        6.0       290.0      17.0   
7   $25,940   $23,508          1

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 428 entries, 0 to 427
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    object 
 6   Invoice      428 non-null    object 
 7   EngineSize   428 non-null    float64
 8   Cylinders    428 non-null    float64
 9   Horsepower   428 non-null    float64
 10  MPG_City     428 non-null    float64
 11  MPG_Highway  428 non-null    float64
 12  Weight       428 non-null    float64
 13  Wheelbase    428 non-null    float64
 14  Length       428 non-null    float64
dtypes: float64(8), object(7)
memory usage: 50.3+ KB


In [16]:
df.isnull().sum()
# so we have not any null values here now

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

## Check what are the different types of Make are there in our dataset. And, what is the count (occurrence) of each Make in the data ?

In [17]:
df.head(5)

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [18]:
df.Make.value_counts()

Toyota           28
Chevrolet        27
Mercedes-Benz    26
Ford             23
BMW              20
Audi             19
Honda            17
Nissan           17
Chrysler         15
Volkswagen       15
Dodge            13
Mitsubishi       13
Jaguar           12
Volvo            12
Hyundai          12
Mazda            11
Pontiac          11
Kia              11
Subaru           11
Lexus            11
Lincoln           9
Mercury           9
Buick             9
GMC               8
Infiniti          8
Cadillac          8
Saturn            8
Suzuki            8
Porsche           7
Saab              7
Acura             7
Oldsmobile        3
Jeep              3
Land Rover        3
Isuzu             2
MINI              2
Scion             2
Hummer            1
Name: Make, dtype: int64

In [19]:
df.Make.unique()

array(['Acura', 'Audi', 'BMW', 'Buick', 'Cadillac', 'Chevrolet',
       'Chrysler', 'Dodge', 'Ford', 'GMC', 'Honda', 'Hummer', 'Hyundai',
       'Infiniti', 'Isuzu', 'Jaguar', 'Jeep', 'Kia', 'Land Rover',
       'Lexus', 'Lincoln', 'MINI', 'Mazda', 'Mercedes-Benz', 'Mercury',
       'Mitsubishi', 'Nissan', 'Oldsmobile', 'Pontiac', 'Porsche', 'Saab',
       'Saturn', 'Scion', 'Subaru', 'Suzuki', 'Toyota', 'Volkswagen',
       'Volvo'], dtype=object)

In [20]:
df.Make.nunique()

38

## - Show all the records where Origin is Asia or Europe.

In [21]:
df.head(5)

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [22]:
df.Origin.value_counts()
# 123+158=    281

Asia      158
USA       147
Europe    123
Name: Origin, dtype: int64

In [23]:
filter_data=df[(df.Origin=='Asia') | (df.Origin=='Europe')]
filter_data

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         28.0   
424  $40,083          2.3        5.0       242.0      20.0         26.0   
425  $42,573          2.9        6.0       268.0      19.0         26.0   
426  $24,641          1.9        4.0       170.0      22.0         29.0   
427  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[281 rows x 15 columns]

##  Remove all the records (rows) where Weight is above 4000.

In [24]:
df.head(5)

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [25]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 428 entries, 0 to 427
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Make         428 non-null    object 
 1   Model        428 non-null    object 
 2   Type         428 non-null    object 
 3   Origin       428 non-null    object 
 4   DriveTrain   428 non-null    object 
 5   MSRP         428 non-null    object 
 6   Invoice      428 non-null    object 
 7   EngineSize   428 non-null    float64
 8   Cylinders    428 non-null    float64
 9   Horsepower   428 non-null    float64
 10  MPG_City     428 non-null    float64
 11  MPG_Highway  428 non-null    float64
 12  Weight       428 non-null    float64
 13  Wheelbase    428 non-null    float64
 14  Length       428 non-null    float64
dtypes: float64(8), object(7)
memory usage: 50.3+ KB


In [26]:
df_w=df[df.Weight>4000]
df_w

Make                           Model    Type  Origin DriveTrain  \
0         Acura                             MDX     SUV    Asia        All   
15         Audi  A4 3.0 Quattro convertible 2dr   Sedan  Europe        All   
17         Audi              A6 4.2 Quattro 4dr   Sedan  Europe        All   
18         Audi                A8 L Quattro 4dr   Sedan  Europe        All   
20         Audi                        RS 6 4dr  Sports  Europe      Front   
..          ...                             ...     ...     ...        ...   
401  Volkswagen                      Touareg V6     SUV  Europe        All   
411  Volkswagen                     Phaeton 4dr   Sedan  Europe      Front   
412  Volkswagen                 Phaeton W12 4dr   Sedan  Europe      Front   
415  Volkswagen                       Passat W8   Wagon  Europe      Front   
416       Volvo                         XC90 T6     SUV  Europe        All   

         MSRP   Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
0    $36,945   $33,337          3.5        6.0       265.0      17.0   
15   $44,240   $40,075          3.0        6.0       220.0      18.0   
17   $49,690   $44,936          4.2        8.0       300.0      17.0   
18   $69,190   $64,740          4.2        8.0       330.0      17.0   
20   $84,600   $76,417          4.2        8.0       450.0      15.0   
..        ...       ...         ...        ...         ...       ...   
401  $35,515   $32,243          3.2        6.0       220.0      15.0   
411  $65,000   $59,912          4.2        8.0       335.0      16.0   
412  $75,000   $69,130          6.0       12.0       420.0      12.0   
415  $40,235   $36,956          4.0        8.0       270.0      18.0   
416  $41,250   $38,851          2.9        6.0       268.0      15.0   

     MPG_Highway  Weight  Wheelbase  Length  
0           23.0  4451.0      106.0   189.0  
15          25.0  4013.0      105.0   180.0  
17          24.0  4024.0      109.0   193.0  
18          24.0  4399.0      121.0   204.0  
20          22.0  4024.0      109.0   191.0  
..           ...     ...        ...     ...  
401         20.0  5086.0      112.0   187.0  
411         22.0  5194.0      118.0   204.0  
412         19.0  5399.0      118.0   204.0  
415         25.0  4067.0      106.0   184.0  
416         20.0  4638.0      113.0   189.0  

[103 rows x 15 columns]

total 428 row removing 103 rows ======= now we will have 325 rows lefy

In [27]:
df2=df[df.Weight<4000]
df2

Make                    Model   Type  Origin DriveTrain      MSRP  \
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
5    Acura  3.5 RL w/Navigation 4dr  Sedan    Asia      Front  $46,100    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
5    $41,100          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         28.0   
424  $40,083          2.3        5.0       242.0      20.0         26.0   
425  $42,573          2.9        6.0       268.0      19.0         26.0   
426  $24,641          1.9        4.0       170.0      22.0         29.0   
427  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
5    3893.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[325 rows x 15 columns]

In [28]:
df2.Weight.max()

3992.0

## Increase all the values of 'MPG_City' column by 3.


In [29]:
df.head(10)

Make                    Model    Type  Origin DriveTrain      MSRP  \
0  Acura                      MDX     SUV    Asia        All  $36,945    
1  Acura           RSX Type S 2dr   Sedan    Asia      Front  $23,820    
2  Acura                  TSX 4dr   Sedan    Asia      Front  $26,990    
3  Acura                   TL 4dr   Sedan    Asia      Front  $33,195    
4  Acura               3.5 RL 4dr   Sedan    Asia      Front  $43,755    
5  Acura  3.5 RL w/Navigation 4dr   Sedan    Asia      Front  $46,100    
6  Acura   NSX coupe 2dr manual S  Sports    Asia       Rear  $89,765    
7   Audi              A4 1.8T 4dr   Sedan  Europe      Front  $25,940    
8   Audi   A41.8T convertible 2dr   Sedan  Europe      Front  $35,940    
9   Audi               A4 3.0 4dr   Sedan  Europe      Front  $31,840    

    Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0  $33,337          3.5        6.0       265.0      17.0         23.0  4451.0   
1  $21,761          2.0        4.0       200.0      24.0         31.0  2778.0   
2  $24,647          2.4        4.0       200.0      22.0         29.0  3230.0   
3  $30,299          3.2        6.0       270.0      20.0         28.0  3575.0   
4  $39,014          3.5        6.0       225.0      18.0         24.0  3880.0   
5  $41,100          3.5        6.0       225.0      18.0         24.0  3893.0   
6  $79,978          3.2        6.0       290.0      17.0         24.0  3153.0   
7  $23,508          1.8        4.0       170.0      22.0         31.0  3252.0   
8  $32,506          1.8        4.0       170.0      23.0         30.0  3638.0   
9  $28,846          3.0        6.0       220.0      20.0         28.0  3462.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0  
5      115.0   197.0  
6      100.0   174.0  
7      104.0   179.0  
8      105.0   180.0  
9      104.0   179.0

In [30]:
df.MPG_Highway


0      23.0
1      31.0
2      29.0
3      28.0
4      24.0
       ... 
423    28.0
424    26.0
425    26.0
426    29.0
427    27.0
Name: MPG_Highway, Length: 428, dtype: float64

In [31]:
df2=df.copy()

In [32]:
df2.MPG_Highway=df2.MPG_Highway+3
df2.MPG_Highway

0      26.0
1      34.0
2      32.0
3      31.0
4      27.0
       ... 
423    31.0
424    29.0
425    29.0
426    32.0
427    30.0
Name: MPG_Highway, Length: 428, dtype: float64

In [33]:
df2
# it incresed the value df.MPG_Highway by adding 3

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         26.0   
1    $21,761          2.0        4.0       200.0      24.0         34.0   
2    $24,647          2.4        4.0       200.0      22.0         32.0   
3    $30,299          3.2        6.0       270.0      20.0         31.0   
4    $39,014          3.5        6.0       225.0      18.0         27.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         31.0   
424  $40,083          2.3        5.0       242.0      20.0         29.0   
425  $42,573          2.9        6.0       268.0      19.0         29.0   
426  $24,641          1.9        4.0       170.0      22.0         32.0   
427  $33,112          2.5        5.0       208.0      20.0         30.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[428 rows x 15 columns]

In [34]:
df
# our orignal data_set is ok

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         28.0   
424  $40,083          2.3        5.0       242.0      20.0         26.0   
425  $42,573          2.9        6.0       268.0      19.0         26.0   
426  $24,641          1.9        4.0       170.0      22.0         29.0   
427  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[428 rows x 15 columns]

In [35]:
df2
# we will again add in this df.MPG_Highway by 3

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         26.0   
1    $21,761          2.0        4.0       200.0      24.0         34.0   
2    $24,647          2.4        4.0       200.0      22.0         32.0   
3    $30,299          3.2        6.0       270.0      20.0         31.0   
4    $39,014          3.5        6.0       225.0      18.0         27.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         31.0   
424  $40,083          2.3        5.0       242.0      20.0         29.0   
425  $42,573          2.9        6.0       268.0      19.0         29.0   
426  $24,641          1.9        4.0       170.0      22.0         32.0   
427  $33,112          2.5        5.0       208.0      20.0         30.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[428 rows x 15 columns]

In [36]:
df3=df2.copy()
df3

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         26.0   
1    $21,761          2.0        4.0       200.0      24.0         34.0   
2    $24,647          2.4        4.0       200.0      22.0         32.0   
3    $30,299          3.2        6.0       270.0      20.0         31.0   
4    $39,014          3.5        6.0       225.0      18.0         27.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         31.0   
424  $40,083          2.3        5.0       242.0      20.0         29.0   
425  $42,573          2.9        6.0       268.0      19.0         29.0   
426  $24,641          1.9        4.0       170.0      22.0         32.0   
427  $33,112          2.5        5.0       208.0      20.0         30.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[428 rows x 15 columns]

In [37]:
# ================================== 2nd method =====================================
df3.MPG_Highway=df3.MPG_Highway.apply(lambda x:x+3)
df3

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
423  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
424  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
425  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
426  Volvo                      V40  Wagon  Europe      Front  $26,135    
427  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         29.0   
1    $21,761          2.0        4.0       200.0      24.0         37.0   
2    $24,647          2.4        4.0       200.0      22.0         35.0   
3    $30,299          3.2        6.0       270.0      20.0         34.0   
4    $39,014          3.5        6.0       225.0      18.0         30.0   
..        ...         ...        ...         ...       ...          ...   
423  $38,203          2.4        5.0       197.0      21.0         34.0   
424  $40,083          2.3        5.0       242.0      20.0         32.0   
425  $42,573          2.9        6.0       268.0      19.0         32.0   
426  $24,641          1.9        4.0       170.0      22.0         35.0   
427  $33,112          2.5        5.0       208.0      20.0         33.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
423  3450.0      105.0   186.0  
424  3450.0      105.0   186.0  
425  3653.0      110.0   190.0  
426  2822.0      101.0   180.0  
427  3823.0      109.0   186.0  

[428 rows x 15 columns]